pyomo/network/plugins/expand_arcs.py

```diff
@@ -0,0 +1,254 @@
+                        "but %s elements in port '%s')" %
+                        (k, v[1], v[2].name, _len, p.name))
+                if v[1] >= 0 and len(v[0].index_set() ^ _v.index_set()):
+                    raise ValueError(
+                        "Port mismatch: Port variable '%s' has "
+                        "mismatched indices on ports '%s' and '%s'" %
+                        (k, v[2].name, p.name))
+                if p._rules[k][0] is not v[3]:
+                    raise ValueError(
+                        "Port mismatch: Port variable '%s' has "
+                        "different rules on ports '%s' and '%s'" %
+                        (k, v[2].name, p.name))
+
+        # as we are adding things to the model, sort by key so that
+        # the order things are added is deterministic
+        sorted_refs = sorted(iteritems(ref))
+        if len(empty_or_partial) > 1:
+            # This is expensive (names aren't cheap), but does result in
+            # a deterministic ordering
+            empty_or_partial.sort(key=lambda x: x.getname(
+                fully_qualified=True, name_buffer=self._name_buffer))
+
+        # Fill in any empty ports
+        for p in empty_or_partial:
+            block = p.parent_block()
+            for k, v in sorted_refs:
+                if k in p.vars and p.vars[k] is not None:
+                    continue
+
+                var_args = {}
+                try:
+                    var_args['domain'] = v[0].domain
+                except AttributeError:
+                    pass
+                try:
+                    var_args['bounds'] = v[0].bounds
+                except AttributeError:
+                    pass
+
+                if v[1] >= 0:
+                    new_var = Var(v[0].index_set(), **var_args)
+                else:
+                    new_var = Var(**var_args)
+                vname = unique_component_name(
+                    block, '%s_auto_%s' % (p.getname(
+                        fully_qualified=True, name_buffer=self._name_buffer), k))
+                block.add_component(vname, new_var)
+                if v[1] >= 0:
+                    for i in v[0].index_set():
+                        # set bounds for every member in case they differ
+                        new_var[i].domain = v[0][i].domain
+                        new_var[i].setlb(v[0][i].lb)
+                        new_var[i].setub(v[0][i].ub)
+                # add this new variable to the port so that it has a rule
+                p.add(new_var, k, rule=v[3])
+
+        return ref
+
+    def _add_blocks(self, instance):
+        # iterate over component_objects so we can make indexed blocks
+        for arc in instance.component_objects(**obj_iter_kwds):
+            blk = Block(arc.index_set())
+            bname = unique_component_name(
+                arc.parent_block(), "%s_expanded" % arc.getname(
+                    fully_qualified=False, name_buffer=self._name_buffer))
```

Comment:
Is this more efficient than using `.local_name`?

Reply:
I think I might have just added this call to getname here as well without thinking when I was speeding up all the name calls. From looking at local_name the only difference seems to be that local_name doesn't use a name buffer, which means this call to getname has the opportunity to return slightly sooner (by a few lines and some function calls). However, considering #608 which I just opened, I may want to switch to local name at least until the issue is resolved.